### PR TITLE
Dev, Tests, Examples and Docs for Entity Groups

### PR DESCRIPTION
**Pull request by abhinavbansal29**
## Entity Groups Modules for Flow Network Security

This PR adds support for managing Entity Groups in Nutanix Flow using v4 APIs.

#### `ntnx_entity_groups_v2`
- Create, update, and delete entity groups
- Support for allowed and except entity configurations
- Entity types: VMs, Subnets, VPCs, Kubernetes resources (namespaces, services, pods, clusters), and Address Groups
- Selection methods: by external ID, category, labels, name, or IP values
- Support for IP addresses (IPv4/IPv6) and IP ranges
- Check mode and idempotency support

#### `ntnx_entity_groups_info_v2`
- Fetch specific entity group by external ID
- List all entity groups with standard v4 list parameters support
- Filtering, pagination, and sorting capabilities

### Key Features
- **Entity Groups** are logical collections of entities that can be referenced in Flow network security policies for microsegmentation
- Support for both allowed entities (what to include) and except entities (what to exclude)
- Comprehensive entity selection options including categories, labels, and IP-based selection
- Full CRUD operations with proper error handling and validation

### Additional Changes
- Updated Flow API client utilities and helpers
- Added constants for entity group operations
- Comprehensive integration tests covering all operations
- Example playbooks demonstrating common use cases
- Complete module documentation

### Testing
- ✅ Integration tests for create, update, delete operations
- ✅ Tests for allowed_config and except_config handling
- ✅ Tests for various entity types and selection methods
- ✅ Check mode and idempotency validation
- ✅ Error handling and edge cases

### CODE REVIEW

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+    description: "{{ eg2 }}_entity_group_desc"
+    allowed_config:
+      entities:
+        - select_by: CATEGORY_EXT_ID
+          type: SUBNET
+          reference_ext_ids:
+            - "{{ todelete_categories[0] }}"
+            - "{{ todelete_categories[1] }}"
+  register: result
+  ignore_errors: true
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]: Group VM and SUBNET in one entity group.

File: examples/flow_management_v2/entity_groups_v2.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        description: "{{ entity_group_description }}_updated_desc"
+      register: result
+      ignore_errors: true
+
+    - name: List all entity groups
+      nutanix.ncp.ntnx_entity_groups_info_v2:
+      register: result
+      ignore_errors: true
+
+    - name: List entity group using filter
```
Comment (george-ghawali) [inline examples/flow_management_v2/entity_groups_v2.yml]:
```suggestion
    - name: List entity groups using filter
```

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+__metaclass__ = type
+
+DOCUMENTATION = r"""
+---
+module: ntnx_entity_groups_v2
+short_description: Create, Update, Delete entity groups
+version_added: 2.5.0
+description:
+    - An Entity Group is a logical group of entities supported by flow network security policies.
+    - Once created, an Entity Group can be referenced in the network security policies.
```
Comment (george-ghawali) [inline plugins/modules/ntnx_entity_groups_v2.py]: can you please elaborate more on how to use these entity groups using select_by and type ? and can we use different types in the same entity group ?

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+    allowed_config:
+      entities:
+        - select_by: CATEGORY_EXT_ID
+          type: VM
+          reference_ext_ids:
+            - "00062ffc-95ad-19e9-185b-ac1f6b6f97a3"
+            - "00062ffc-95ad-19e9-185b-ac1f6b6f97a4"
+  register: result
+  ignore_errors: true
+
```
Comment (george-ghawali) [inline plugins/modules/ntnx_entity_groups_v2.py]: please add an example for kube_entites, addresses, and ip_ranges

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+  register: result
+  ignore_errors: true
+"""
+
+RETURN = r"""
+response:
+  description:
+      - Response for entity groups operations
+      - Entity group details if C(wait) is True
+      - Task details if C(wait) is False
```
Comment (george-ghawali) [inline plugins/modules/ntnx_entity_groups_v2.py]: we can remove this line if we add the previous comment

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+
+    module_args = dict(
+        ext_id=dict(type="str"),
+        name=dict(type="str"),
+        description=dict(type="str"),
+        allowed_config=dict(
+            type="dict",
+            options=allowed_config_spec,
+            obj=mic_sdk.AllowedConfig,
+        ),
```
Comment (george-ghawali) [inline plugins/modules/ntnx_entity_groups_v2.py]: pleae add missing attribute except_config

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+            exception=e,
+            msg="Api Exception raised while updating entity_group",
+        )
+
+    task_ext_id = resp.data.ext_id
+    result["task_ext_id"] = task_ext_id
+    result["response"] = strip_internal_attributes(resp.data.to_dict())
+
+    if task_ext_id and module.params.get("wait"):
+        wait_for_completion(module, task_ext_id, True)
```
Comment (george-ghawali) [inline plugins/modules/ntnx_entity_groups_v2.py]:
```suggestion
        wait_for_completion(module, task_ext_id)
```

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+    ext_id = module.params.get("ext_id")
+    result["ext_id"] = ext_id
+
+    if module.check_mode:
+        result["msg"] = "Entity group with ext_id:{0} will be deleted.".format(ext_id)
+        return
+
+    current_spec = get_entity_group(module, entity_group, ext_id)
+
+    etag = get_etag(data=current_spec)
```
Comment (george-ghawali) [inline plugins/modules/ntnx_entity_groups_v2.py]: no etag needed for delete based on API docs

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+
+
+def update_entity_group(module, entity_group, result):
+    ext_id = module.params.get("ext_id")
+    result["ext_id"] = ext_id
+
+    current_spec = get_entity_group(module, entity_group, ext_id)
+    sg = SpecGenerator(module)
+    update_spec, err = sg.generate_spec(obj=deepcopy(current_spec))
+
```
Comment (george-ghawali) [inline plugins/modules/ntnx_entity_groups_v2.py]: etag is needed for update based on API docs

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+            exception=e,
+            msg="Api Exception raised while deleting entity_group",
+        )
+
+    task_ext_id = resp.data.ext_id
+    result["task_ext_id"] = task_ext_id
+    result["response"] = strip_internal_attributes(resp.data.to_dict())
+
+    if task_ext_id and module.params.get("wait"):
+        resp = wait_for_completion(module, task_ext_id, True)
```
Comment (george-ghawali) [inline plugins/modules/ntnx_entity_groups_v2.py]:
```suggestion
        resp = wait_for_completion(module, task_ext_id)
```

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+- name: Set todelete list
+  ansible.builtin.set_fact:
+    todelete: []
+
+- name: Set todelete_categories list
+  ansible.builtin.set_fact:
+    todelete_categories: []
+
+- name: Create entity group with check mode enabled
+  nutanix.ncp.ntnx_entity_groups_v2:
```
Comment (george-ghawali) [inline tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]: please use all attributes

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 12]
```diff
@@ -0,0 +1,518 @@
...
+    eg2: "{{ random_name }}{{ suffix_name }}2"
+
+- name: Set ansible keys and values
+  ansible.builtin.set_fact:
+    keys:
+      - "{{random_name}}{{suffix_name}}key1"
+      - "{{random_name}}{{suffix_name}}key2"
+      - "{{random_name}}{{suffix_name}}key3"
+      - "{{random_name}}{{suffix_name}}key4"
+    values:
```
Comment (george-ghawali) [inline tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]: we only used 2 in the test, please remove extra vars

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 13]
```diff
@@ -0,0 +1,518 @@
...
+      - result.response is defined
+      - result.changed == False
+      - result.failed == True
+      - result.msg == "Task Failed"
+    fail_msg: Create entity group with wrong reference_ext_ids did not fail as expected
+    success_msg: Create entity group with wrong reference_ext_ids failed as expected
+
+###################################################################################################
+
+- name: Update entity group in check mode
```
Comment (george-ghawali) [inline tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]:
```suggestion
- name: Generate spec for updating entity group using check mode
```

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 14]
```diff
@@ -0,0 +1,518 @@
...
+      - result.response.allowed_config.entities[0].select_by == "CATEGORY_EXT_ID"
+      - result.response.allowed_config.entities[0].type == "SUBNET"
+      - result.response.allowed_config.entities[0].reference_ext_ids == ["00062ffc-95ad-19e9-185b-ac1f6b6f97e2", "00062ffc-95ad-19e9-185b-ac1f6b6f97a3"]
+      - result.response.allowed_config.entities[0].reference_ext_ids | length == 2
+    fail_msg: Update entity group in check mode failed
+    success_msg: Update entity group in check mode passed
+
+###################################################################################################
+
+- name: Update entity group
```
Comment (george-ghawali) [inline tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]: Please add more attributes

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 15]
```diff
@@ -0,0 +1,518 @@
...
+      - result.failed == False
+      - result.response.name == "{{ eg2 }}_entity_group"
+      - result.response.ext_id == "{{ todelete[1] }}"
+      - result.response.description == "{{ eg2 }}_entity_group_desc"
+      - result.response.allowed_config.entities[0].select_by == "CATEGORY_EXT_ID"
+      - result.response.allowed_config.entities[0].type == "SUBNET"
+      - result.response.allowed_config.entities[0].reference_ext_ids == todelete_categories
+      - result.response.allowed_config.entities[0].reference_ext_ids | length == todelete_categories | length
+    fail_msg: Get entity group using ext_id failed
+    success_msg: Entity group listed successfully using ext_id
```
Comment (george-ghawali) [inline tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]:
```suggestion
    success_msg: Get entity group using ext_id passed successfully
```

File: plugins/modules/ntnx_entity_groups_v2.py [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+                    - List of allowed entities in the Entity Group.
+                    - An allowed entity is a collection of acceptable entites.
+                type: list
+                elements: dict
+                suboptions:
+                    select_by:
+                        description:
+                            - Select by field for the allowed entity.
+                        type: str
+                        choices: ["EXT_ID", "CATEGORY_EXT_ID", "LABELS", "NAME"]
```
[the remark this answers is not in the document]
Comment (abhinavbansal29) [reply plugins/modules/ntnx_entity_groups_v2.py]: Added now, It was recently added in SDKs as part of v4.2.

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 17]
```diff
@@ -0,0 +1,518 @@
...
+    that:
+      - result.response is defined
+      - result.changed == True
+      - result.failed == False
+      - result.ext_id is defined
+      - result.response.name == "{{ eg1 }}_entity_group"
+      - result.response.description == "{{ eg1 }}_entity_group_desc"
+      - result.response.allowed_config.entities[0].select_by == "CATEGORY_EXT_ID"
+      - result.response.allowed_config.entities[0].type == "VM"
+      - result.response.allowed_config.entities[0].reference_ext_ids == todelete_categories
```
[the remark this answers is not in the document]
Comment (abhinavbansal29) [reply tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]: We just have one entity.

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 18]
```diff
@@ -0,0 +1,518 @@
...
+          reference_ext_ids:
+            - "83cbf00d-782f-4efc-87c8-4129f5942aaa"
+            - "83cbf00d-782f-4efc-87c8-4129f5942bbb"
+  register: result
+  ignore_errors: true
+
+- name: Create entity group with wrong reference_ext_ids Status
+  ansible.builtin.assert:
+    that:
+      - result.response is defined
```
[the remark this answers is not in the document]
Comment (abhinavbansal29) [reply tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]: These are the only fields which we get, so can't add more assertions.

File: tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml [hunk 19]
```diff
@@ -0,0 +1,518 @@
...
+  loop: "{{ range(0, 2) | list }}"
+  loop_control:
+    label: "{{ item }}"
+
+- name: Create categories for entity groups status
+  ansible.builtin.assert:
+    that:
+      - output is defined
+      - output.changed == true
+      - output.results | length == 2
```
[the remark this answers is not in the document]
Comment (abhinavbansal29) [reply tests/integration/targets/ntnx_entity_groups_v2/tasks/entity_groups.yml]: We can't assert for failed here.

**Comment by abhinavbansal29** [comment]
/ok-to-test ntnx_entity_groups_v2
> Integration test run status is : failure
> Coverage is : 41 %
> Job link: https://github.com/nutanix/nutanix.ansible/actions/runs/22139747346

### Detailed Test Results
```text
### Summary:
- Total Tests: 1
- Passed: 0
- Failed: 1

### Passed Tests:
- None

### Failed Tests:
- ntnx_entity_groups_v2

Raw log file: test_output.log
```

**Comment by abhinavbansal29** [comment]
/ok-to-test ntnx_entity_groups_v2
> Integration test run status is : failure
> Authentication method: 🔐 Username/Password
> Coverage is :  %
> Job link: https://github.com/nutanix/nutanix.ansible/actions/runs/22609314075

### Detailed Test Results
```text
No detailed results (test_output.log missing)
```

**Comment by abhinavbansal29** [comment]
/ok-to-test ntnx_entity_groups_v2
> Integration test run status is : success
> Authentication method: 🔐 Username/Password
> Coverage is : 62 %
> Job link: https://github.com/nutanix/nutanix.ansible/actions/runs/22610093834

### Detailed Test Results
```text
### Summary:
- Total Tests: 1
- Passed: 1
- Failed: 0

### Passed Tests:
- ntnx_entity_groups_v2

### Failed Tests:
- None

Raw log file: test_output.log
```

**Review by george-ghawali — APPROVED** [review]
